### ISSUE TITLE
re-enable erasure for broadcast and repair

### ISSUE THREAD

**Issue by rob-solana** [solana-labs/solana#636]
Current erasure code introduces blob index skew from entry height, which is a problem if leaders and validators start with different initial entry height.

Erasure is therefore disabled until an "infinite" repair window can be supported, via a persistent, random-access ledger.  Erasure coding can then run on a broadcast/repair specific numbering scheme.

https://github.com/solana-labs/solana/pull/637 should be reverted as a part of resolving this issue

**Comment by aeyakovenko** [comment]
You could transmit erasure blobs with the same index.  They have a bool set in the blob header already

**Comment by garious** [comment]
@aeyakovenko, no pronouns please. I don't understand your comment.

**Comment by rob-solana** [comment]
I don't think using the entry height-denoting value of blob_index twice (or N times) will work, but I'd have to go read the code more closely.  @sakridge do we have a design doc for how our erasure coding of the window blobs is supposed to work?

**Comment by rob-solana** [comment]
The plan:
- [ ] Window, which today is roughly Vec<Option<SharedBlob>>, will become Vec<(Option<SharedBlob>,Option<SharedBlob>)>...  The tuple will allow an erasure coding Blob to co-exist with an actual data SharedBlob for broadcast, receive, repair in the same Window Slot (perhaps a better name for the type?)
- [ ] erasure, instead of adding Blobs to the Window by lengthening (and thereby skewing blob_idx), will put the coding Blobs alongside the data Blobs in the Window (the spread of MAX_MISSING over NUM_CODED TBD, probably don't need to be any more clever than the current "append" semantics...)
- [ ] both data Blobs and coding Blobs will be broadcast, sent over repair, etc.

**Comment by aeyakovenko** [comment]
@sakridge, robs plan seems pretty good to me

**Comment by garious** [comment]
I don't quite grok it. `(Option, Option)` is a type with 4 possible values. Would all 4 be utilized here?

**Comment by aeyakovenko** [comment]
Yes, not all the indexied have a corresponding erasure blob.  Erasure rate could be set to any rate by the leader. The validator may not have received a data blob for any index.

**Comment by garious** [comment]
So `(Blob, Option<Blob>)`?

**Comment by rob-solana** [comment]
we need:

data, coded
None, none: initial state
None, Some: only have coded
Some, None: talking to a non-coded peer, or have received a data and a coded
Some, Some: talking to a coded peer, but possibly missing an earlier data Blob

**Comment by garious** [comment]
Nice, thanks for the detail!

**Comment by sakridge** [comment]
@aeyakovenko why not repair erasure blobs, that may be the fastest way to recover a coded blob set?

@rob-solana Don't really have any design doc, just the code. Plan looks good. I didn't want to mess with the repair logic I think that's why I kept the coded blobs at unique indices. One way to fix the skew is to have maybe special 'coded' entries in the log file, am I understanding it right?

As for design doc, general idea is that it inserts the coding blobs into the window at a fixed coding rate. On the generate side, when we reach the 'end' of the data blob set, then we insert the coding blobs before broadcasting them to peers. That's in streamer::broadcast which calls erasure::add_coding_blobs. These blobs are like the data blobs but there is a coding bit set, and the coding relies on the data blobs being the same size, so the coding is the maximum size of all data blobs in the set and the data is assumed to be 0-padded to that size to be able to restore the set.

On the consume side, in streamer::process_blob(), we need to skip over coding blobs for processing, so we jump over their consumed values when sending the blob to the banking stage. There is some logic in there to keep the data blobs around for longer than we used to be doing in the original code in which the non-erasure path was null-ing out the window. We now keep the blobs around in the non-erasure path to be able to respond to peer repair requests, so the erasure side could be updated to match that.

After receiving new blobs, we try to recover a blob set in streamer::repair_window() with the erasure::recover function which looks over the window and sees where it can repair a blob set in the window. It needs to be only missing the number of coding blobs in a set.

Anyway, let me know if you have more questions.

> both data Blobs and coding Blobs will be broadcast, sent over repair, etc.
For a repair request, I guess a peer would respond with both data/coding if it had it? Would it distinguish between the case where it only needed one or the other?
